app/src/main/java/com/kickstarter/viewmodels/SurveyResponseViewModel.java

```diff
@@ -0,0 +1,155 @@
+import rx.subjects.PublishSubject;
+
+import static com.kickstarter.libs.rx.transformers.Transformers.ignoreValues;
+import static com.kickstarter.libs.rx.transformers.Transformers.neverError;
+
+public interface SurveyResponseViewModel {
+
+  interface Inputs {
+    /** Call when a project uri request has been made. */
+    void projectUriRequest(Request request);
+
+    /** Call when the dialog's OK button has been clicked. */
+    void okButtonClicked();
+
+    /** Call when a project survey uri request has been made. */
+    void projectSurveyUriRequest(Request request);
+
+    /** Call when a web view page has been intercepted. */
+    void webViewPageIntercepted(String url);
+  }
+
+  interface Outputs {
+    /** Emits when we should navigate back. */
+    Observable<Void> goBack();
+
+    /** Emits when we should show a confirmation dialog. */
+    Observable<Void> showConfirmationDialog();
+
+    /** Emits a project and a ref tag to start the {@link com.kickstarter.ui.activities.ProjectActivity} with. */
+    Observable<Pair<Project, RefTag>> startProjectActivity();
+
+    /** Emits a url to load in the web view. */
+    Observable<String> webViewUrl();
+  }
+
+  final class ViewModel extends ActivityViewModel<SurveyResponseActivity> implements Inputs, Outputs {
+    private final ApiClientType client;
+
+    public ViewModel(final @NonNull Environment environment) {
+      super(environment);
+
+      this.client = environment.apiClient();
+
+      final Observable<SurveyResponse> surveyResponse = intent()
+        .map(i -> i.getParcelableExtra(IntentKey.SURVEY_RESPONSE))
+        .ofType(SurveyResponse.class);
+
+      final Observable<String> surveyWebUrl = surveyResponse
+        .map(r -> r.urls().web().survey());
+
+      surveyWebUrl
+        .compose(bindToLifecycle())
+        .subscribe(this.webViewUrl);
+
+      final Observable<Pair<Request, String>> projectRequestAndSurveyUrl = Observable.zip(
+        this.projectUriRequest,
+        surveyWebUrl,
+        Pair::create
+      );
+
+      final Observable<Project> project = projectRequestAndSurveyUrl
+        .filter(requestAndUrl -> !requestTagUrlIsSurveyUrl(requestAndUrl))
+        .map(PairUtils::first)
+        .map(this::extractProjectParams)
+        .switchMap(this.client::fetchProject)
+        .compose(neverError())
+        .share();
+
+      projectRequestAndSurveyUrl
+        .filter(this::requestTagUrlIsSurveyUrl)
+        .compose(ignoreValues())
+        .compose(bindToLifecycle())
+        .subscribe(this.showConfirmationDialog);
+
+      project
+        .compose(bindToLifecycle())
+        .subscribe(p -> this.startProjectActivity.onNext(Pair.create(p, RefTag.survey())));
+
+      this.goBack = this.okButtonClicked;
+    }
+
+    /**
+     * Parses a project request for project params.
+     */
+    private @NonNull String extractProjectParams(final @NonNull Request request) {
+      return request.url().encodedPathSegments().get(2);
```

Comment:
you probably already thought about this, but could the list returned by `encodedPathSegments` ever be `null` or a length shorter than two?

Reply:
adding idk if there is anything to be done about it even if there is.

Reply:
you are quite right, this definitely is not an ideal way to handle parsing project params. fwiw it's a todo on our UpdateViewModel as well, to build a Navigation helper for safer extraction that i'd like to make eventually. for now it's ok because our request handler pretty much guarantees that a project request is coming through but seeing this `.get(2)` is 😨